### ISSUE TITLE
[RFC]: Async Failure Notification for Fault Tolerant EP Kernels

### ISSUE THREAD

**Issue by tzulingk** [vllm-project/vllm#39760]
# [RFC]: Async Failure Notification for Fault Tolerant EP Kernels

### Motivation.

Fault-tolerant EP kernels (e.g. NIXL-EP) detect rank failures on the GPU — via a per-rank
integer mask written atomically on timeout — without any CPU involvement. The CPU only
learns about a failure if it explicitly reads that mask back from VRAM, which forces a
GPU→CPU sync.

The problem is not detection speed. The problem is **when and how often that sync is paid**.

When [`SchedulerConfig.async_scheduling = True`](https://github.com/vllm-project/vllm/blob/54b0578adacd0413b18c0f59948dabc7533a6524/vllm/config/scheduler.py#L144), vLLM pipelines CPU scheduling with GPU
execution: the CPU schedules pass N+1 while the GPU runs pass N. A blocking GPU→CPU sync
collapses that overlap and eliminates the throughput benefit of async scheduling — even
during healthy operation when no rank has failed.

The natural "check after every forward pass" approach pays this cost on every pass:

```python
query_mask_buffer(...)     # enqueues a CUDA copy kernel
active_ranks.copy_(...)    # still a GPU tensor
torch.equal(active_ranks, ...)  # blocks — drains the CUDA stream, crosses PCIe
```

As the NIXL team noted:
> "Raising a formal exception requires reading the failure mask back to the CPU.
> This synchronization point may introduce the very latency we are trying to avoid
> by using async runs."

This RFC proposes wiring GPU-side mask detection into vLLM in a way that pays the sync
cost **only when a failure actually occurs**, not on every pass.

---

### Proposed Change.

Two designs are proposed.

---

#### Design 1 — Pinned Dirty Flag

The GPU kernel writes a single integer — `dirty_flag` — in **pinned (page-locked) CPU
memory** whenever a rank fails. Pinned memory has a fixed physical address, so the GPU
can write to it directly over PCIe without any CPU involvement or staging. The CPU reads
it for free since it is already in its own RAM.

One store is added to the EP dispatch/combine kernels alongside the existing `atomicExch`:

```cpp
// In dispatch / combine kernel, when timeout occurs:
atomicExch(mask_buffer_ptr + src_rank, 1);    // existing: per-rank mask in VRAM
st_release_sys_global(dirty_flag_pinned, 1);  // new: single bit written to CPU RAM
```

##### Where to Check the Dirty Flag

The check can be placed at two points. The placement determines detection lag and how
much compute is wasted on failure:

```
Option A — inside All2AllManagerBase.dispatch() [recommended]:

  MoE layer N:    dispatch → experts → combine  ← failure detected
  MoE layer N+1:  dispatch ← check here → raise EPRankFailureError
                             stops here; remaining layers never run

Option B — after model.forward() returns:

  MoE layer N:    dispatch → experts → combine  ← failure detected
  MoE layer N+1:  dispatch → experts → combine  (runs anyway, wasted)
  ...
  model.forward() returns → check dirty_flag → handle
```

| | Option A: check in `dispatch()` | Option B: check at pass boundary |
|---|---|---|
| Detection lag | ≤ one MoE layer | Remainder of current pass |
| Wasted compute on failure | One MoE layer | All remaining MoE layers |
| Exception call stack | Natural unwind to `execute_model()` | No active call stack |

Option A is strictly better on all three dimensions.

##### Hot-Path Cost

The check runs once per MoE layer. Accessing the flag via `ctypes` avoids Python tensor
boxing overhead and keeps the per-check cost minimal — a direct read from CPU RAM:

```python
import ctypes
dirty_flag_raw = ctypes.cast(dirty_flag.data_ptr(), ctypes.POINTER(ctypes.c_int32))
# dirty_flag_raw[0] — direct C memory access, no Python boxing
```

`ctypes` is significantly cheaper than PyTorch tensor indexing (`dirty_flag[0]`), which
incurs Python object overhead on every call. Since failures are rare, the branch predictor
predicts "not taken" almost perfectly. The misprediction cost is paid only on the one
check that actually fires.

##### Exception Flow

When `dirty_flag_raw[0] != 0`, the handler does a **one-time stream sync** to drain the
GPU, reads the full mask to identify failed ranks, resets the flag, and raises
`EPRankFailureError`. The exception unwinds the call stack naturally:

```
All2AllManagerBase.dispatch()     ← raises EPRankFailureError
  GroupCoordinator.dispatch()
    DefaultMoERunner._maybe_dispatch()
      model.forward()
        model_runner.execute_model()
          worker.execute_model()  ← catches; self.elastic_ep_executor is here
```

```python
# All2AllManagerBase.dispatch() — called once per MoE layer
def dispatch(self, hidden_states, topk_weights, topk_ids, ...):
    if dirty_flag_raw[0] != 0:                     # cheap read from CPU RAM
        torch.cuda.current_stream().synchronize()  # sync once, only on failure
        buffer.query_mask_buffer(mask_cpu)         # stream is drained; read full mask
        failed_ranks = mask_cpu.nonzero().tolist()
        dirty_flag_raw[0] = 0
        raise EPRankFailureError(failed_ranks)
    # ... normal dispatch logic

# worker.execute_model() — the catch boundary
def execute_model(self, scheduler_output):
    try:
        output = self.model_runner.execute_model(scheduler_output, ...)
    except EPRankFailureError as e:
        send_zmq_failure_notification(e.failed_ranks)   # new ZMQ message type
```

The mask readback is deferred to the moment a failure actually happens, paid only once,
not on every dispatch call.

**Properties:**

| | |
|---|---|
| Hot-path cost (healthy) | Negligible — one integer read from CPU RAM (`ctypes`) |
| Detection lag | ≤ one MoE layer |
| Mask readback sync | `stream.synchronize()` once, only on failure |
| Pinned memory required | One `int32` (`dirty_flag`) |
| Exception needed | Yes — to unwind the forward pass call stack |
| C++ required | No |
| GPU kernel change | Yes — one additional store on the timeout path |

---

#### Design 2 — `cudaLaunchHostFunc` + ZMQ Notification

CUDA's `cudaLaunchHostFunc` places a CPU function onto a CUDA stream. It executes on a
CUDA-managed background thread after the preceding GPU work completes, entirely
independent of the main thread:

```
CUDA stream:  [dispatch kernel] → [combine kernel] → [host callback] → [next kernels ...]
Main thread:  submits work ─────────────────────────────────────────→ prepares next batch
Background:                        failure detected    sends ZMQ message → done
```

The main thread submits GPU work and immediately moves to the next batch. The callback
fires in the background, checks the mask, and sends a ZMQ failure notification to the
engine core. No exception is raised — the background thread has no forward pass call
stack to unwind.

##### Pinned Memory Requirement

Design 2 requires a **full pinned `mask_cpu` array** (not just a single integer). The
background CUDA thread cannot call CUDA APIs or touch VRAM, so the mask must already be
in CPU RAM when the callback fires. This is done by queuing `query_mask_buffer` into the
stream before the host callback:

```python
mask_cpu = torch.zeros(num_ranks, dtype=torch.int32).pin_memory()

# After combine kernel, enqueue in order:
buffer.query_mask_buffer(mask_cpu)            # (1) CUDA kernel: VRAM → pinned CPU RAM
cudaLaunchHostFunc(stream, on_combine_done)   # (2) fires after (1) completes
```

PyTorch does not expose `cudaLaunchHostFunc` in Python, so a small C++ extension
is required.

##### Notification Path

vLLM separates GPU workers from the engine core across processes:

```
Engine Core Client (core_client.py)   ←——— ZMQ ———→   GPU Worker Processes
  asyncio event loop                                     execute_model()
  abort_requests_async()                                 dispatch / combine kernels
  _scale_down_elastic_ep()                               mask_buffer_ptr
```

The background CUDA thread cannot reach the engine core's asyncio loop directly.
It sends a ZMQ message, which the engine core receives on its next event loop iteration:

```python
# Worker process — CUDA background thread
def on_combine_done():
    failed_ranks = [i for i in range(num_ranks) if mask_cpu[i] != 0]
    if failed_ranks:
        zmq_socket.send_pyobj(("GPU_MASK_FAILURE", failed_ranks))  # non-blocking

# Engine core client — asyncio event loop, next iteration
async def _handle_gpu_mask_failure(failed_ranks: list[int]):
    await self._abort_requests(inflight_for(failed_ranks))
    await self._scale_down_elastic_ep(
        cur_data_parallel_size=len(self.core_engines),
        new_data_parallel_size=len(self.core_engines) - len(failed_ranks),
        dead_dp_ranks=failed_ranks,
    )
```

asyncio is cooperative, not preemptive. The background thread signals and returns
immediately; the engine picks up the message at the next iteration boundary after the
forward pass completes. Interrupting mid-pass is intentionally avoided — the GPU kernels
already masked the failed rank and continued, so the current pass output is valid.

**Properties:**

| | |
|---|---|
| Hot-path cost (healthy) | Zero — background thread only |
| Detection lag | Sub-pass (fires when combine kernel finishes) |
| Mask readback sync | None — GPU wrote to pinned CPU RAM in-stream |
| Pinned memory required | Full `mask_cpu` array (`num_ranks × int32`) |
| Exception needed | No — background thread communicates via ZMQ |
| C++ required | Yes (`cudaLaunchHostFunc` binding) |
| GPU kernel change | No — mask is already written |

---

#### Design Comparison

| | Design 1: Dirty Flag | Design 2: `cudaLaunchHostFunc` |
|---|---|---|
| Hot-path overhead (healthy) | Negligible — one integer read from CPU RAM | Zero |
| Detection lag | ≤ one MoE layer | Sub-pass |
| Pinned memory required | One `int32` | Full `mask_cpu` array |
| Mask readback sync | `stream.synchronize()` once on failure | None |
| Exception needed | Yes — unwinds call stack | No — ZMQ sideways |
| C++ required | No | Yes |
| Effort | Low — pure Python + small kernel change | Medium — C++ extension needed |

The per-dispatch overhead of Design 1 is negligible in production. Design 1 is the
pragmatic first step; Design 2 is the preferred end state for zero hot-path cost.

---

#### Recovery Path (Reusing PR [#38862](https://github.com/vllm-project/vllm/pull/38862))

Once the engine core receives a failure notification, the same three recovery actions
from PR #38862 follow:

| Step | Action | Entry point |
|---|---|---|
| 1. Client notification | Send `FinishReason.ABORT` to in-flight requests | `abort_requests_async()` in `core_client.py` |
| 2. DP routing update | Stop routing new requests to the failed engine | `self.core_engines` trim in `_scale_down_elastic_ep()` |
| 3. Expert redistribution | Reassign experts; reload from disk if needed | `elastic_execute.py` → `reassign_missing_experts()` |

No new recovery logic is needed. Both designs plug into the existing machinery as an
**earlier trigger** — GPU mask detection fires within one forward pass, while PR #38862's
process monitor fires only after the OS confirms a process crash (potentially seconds
later).

---

#### Design Decisions

**1. API boundary for `dirty_flag_ptr`**

The dirty flag address must be passed to the kernel buffer at initialization. This should
be an **internal implementation detail** of each `All2AllManagerBase` subclass — not part
of a public kernel buffer API — so different subclasses can adopt the flag independently
without requiring an API contract change.

**2. C++ binding location for Design 2**

The `cudaLaunchHostFunc` binding should live in the kernel buffer library (e.g. as a
`set_failure_callback` API on `nixl_ep.Buffer`) rather than in the vLLM repo. This
avoids duplicating the binding in each downstream consumer.

**3. New exception type: `EPRankFailureError(RuntimeError)`**

A new exception class `EPRankFailureError(RuntimeError)` in
`vllm/distributed/elastic_ep/` should be created, following the pattern of
`_BarrierTimeoutError(RuntimeError)` in
[`elastic_state.py:73`](https://github.com/vllm-project/vllm/blob/54b0578adacd0413b18c0f59948dabc7533a6524/vllm/distributed/elastic_ep/elastic_state.py#L73).
It should carry `failed_ranks: list[int]` to identify which EP slots failed.

**4. Dirty flag reset timing**

The CPU resets `dirty_flag_raw[0] = 0` after reading the full `mask_buffer_ptr`, before
`reconnect_peer` completes. This is safe because `connect_ranks` clears the GPU-side
`mask_buffer_ptr` entries for reconnected ranks — the next pass will see a clean mask for
the recovered slot. A CPU-side reset is sufficient; no kernel change is needed for reset.

**5. Multiple simultaneous failures**

A single `dirty_flag` integer coalesces multiple failures in one pass into a single
detection event. This is acceptable because the one-time stream sync that follows reads
the full `mask_buffer_ptr` to enumerate all failed rank indices. The recovery path in
PR #38862 already handles a list of failed ranks.

### CC List.

@itayalroy @fangyuchu @tylermsmith @sagemoore

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by fangyuchu** [comment]
Hi @tzulingk , I have a question about Design 2.

If failure detection and notification happen asynchronously on a separate thread, is there any guarantee that the token being generated at the time of failure won’t still be returned to the engine core?

In particular, if a rank fails during the current step (not sure if it is still possible to continue computing with partial ranks), which would make the token incorrect because of the missing expert contributions. In that case, could such a token still be propagated back and treated as a valid output before the failure signal is processed?

**Comment by SageMoore** [comment]
Hi, @tzulingk. Thanks for the detailed design overview! I'm worried that Design 1 Option A won't work with CUDA graphs, since `All2AllManagerBase.dispatch()` isn't called during replay.
